Milestone 1.1 (Helping learners when they get Stuck): State migration for the field `dest_if_really_stuck`

Manan-Rathi opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #[fill_in_number_here].
2. This PR does the following: Adds the dest_if_really_stuck field to the Outcome class and in another derived field default_outcome. This field would help in creation of an alternate dest to direct the learners to, if they are really stuck.

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

If the changes in your PRs are autogenerated via a script and you cannot provide proof
for the changes then please leave a comment "No proof of changes needed because {{Reason}}"
and remove all the sections below.


-->

https://user-images.githubusercontent.com/76519771/177944292-154e496b-842b-4144-bcea-f74789a04894.mp4


Steps followed: 
1. Started dev server in develop using flag  --save_datastore
2. Created a new exploration on develop, and saved it.
3. Downloaded the YAML, no dest_if_stuck field existed.
4. Stopped the server.
5. Checkout to migration branch, and start dev server using same flag.
6. Opened the previously created new exploration.
7. New field exists in the updated YAML file.


#### Proof of changes on desktop with slow/throttled network
Not required since no frontend changes are made, only a new field is added to Outcome and DefaultOutcome
<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI in any of the files listed in rtl_css.py (i.e, those that have
a separate .rtl.css file for styling), make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

Manan-Rathi committed:
Initial changes.

Manan-Rathi committed:
Added for dest_if_really_stuck.

Manan-Rathi committed:
Added catchMisspellings.

Manan-Rathi committed:
Added tests

Manan-Rathi committed:
primary fixes

Manan-Rathi committed:
lint fixes

Manan-Rathi committed:
ts-fix

Manan-Rathi committed:
ft-fixes

Manan-Rathi committed:
ft-fixes

Manan-Rathi committed:
fix

Manan-Rathi committed:
ft-fixes

Manan-Rathi committed:
minor fix

Manan-Rathi committed:
Started resolving merge-conflicts

Manan-Rathi committed:
Merge conflicts

Manan-Rathi committed:
Lint fix

Manan-Rathi committed:
ts chks fix

Manan-Rathi committed:
fixes

Manan-Rathi committed:
changes

Manan-Rathi committed:
Trying to fix e2e fails.

Manan-Rathi committed:
Trying to fix e2e fails-2.

Manan-Rathi committed:
Trying

Manan-Rathi committed:
historyjs changed

Manan-Rathi committed:
c

Manan-Rathi committed:
Revert "c"

This reverts commit 1f3fb07b6b4ae6fcbc67e02ab620b0e2b601a465.

Manan-Rathi committed:
Revert "Revert "c""

This reverts commit 0bae462e7825d58615f4a8fb1161d979bb1db366.

Manan-Rathi committed:
Reverted several commits

Manan-Rathi committed:
fix ts-errs

Manan-Rathi committed:
Add c_m field to several tests.

Manan-Rathi committed:
Changed HistoryTab.js

Manan-Rathi committed:
Some additions

Manan-Rathi committed:
Changes in draft + change in question_domain if condition.

Manan-Rathi committed:
Changes in draft,exp_d,exp_s,backend_t.

Manan-Rathi committed:
lintfix

Manan-Rathi committed:
Fix backend tests.

Manan-Rathi committed:
Removed catchMisspellings

Manan-Rathi committed:
Resolve historyTab & draft issues

Manan-Rathi committed:
Changes desc of convert func

Manan-Rathi committed:
Trying fixing history.js

Manan-Rathi committed:
Backend error fixed

Manan-Rathi committed:
H-t-1

Manan-Rathi committed:
H-t-2

Manan-Rathi committed:
Desc changed

Manan-Rathi committed:
Merge branch 'develop' of https://github.com/oppia/oppia into gsoc-sp1.1

Manan-Rathi committed:
Edits in hitsory and backend fixes

Manan-Rathi committed:
fix lint

Manan-Rathi committed:
fix  back more

Manan-Rathi committed:
fix  back more

Manan-Rathi committed:
fix  back-2

Manan-Rathi committed:
Fix conflicts

Manan-Rathi committed:
ts-fixes

Manan-Rathi committed:
lint fix

Manan-Rathi committed:
Trying fixing historytb

Manan-Rathi committed:
add test

Manan-Rathi committed:
lint fix

Manan-Rathi committed:
Try-1

Manan-Rathi committed:
try-2

Manan-Rathi committed:
try-3

Manan-Rathi committed:
Fixes

Manan-Rathi commented:
PTAL @vojtechjelinek @kevintab95 @gp201 @krishnarao22

vojtechjelinek commented:
@Manan-Rathi Your mentor should review this first.

iamprayush commented on core/domain/caching_services_test.py:
```diff
@@ -164,30 +166,31 @@ class CachingServicesUnitTests(test_utils.GenericTestBase):
-        '}}, "version": 0, "correctness_feedback_enabled": false, "edits_allowe'
-        'd": true, "language_code": "en", "objective": "", "init_state_name": "'
-        'Introduction", "blurb": "", "author_notes": ""}'
+        's great! \\u00ae\\u00ae</p>"}, "dest": "Introduction", "dest_if_really'
```
If you're making a change here, don't you also need to update the states_schema_version? Also, does this not go through the migration functions when it's loaded? Why do you need to explicitly add the field here?

Manan-Rathi replied:
Right, shouldn't have been there, fixed.

iamprayush commented on core/domain/draft_upgrade_services.py:
```diff
@@ -207,6 +207,23 @@ def _convert_html_in_draft_change_list(
 
+    @classmethod
+    def _convert_states_v50_dict_to_v51_dict(cls, draft_change_list):
+        """Converts draft change list from state version 49 to 50. Version 50
```
*50 to 51

Manan-Rathi replied:
Fixed.

iamprayush commented on core/domain/exp_domain.py:
```diff
@@ -2191,6 +2191,30 @@ def _convert_states_v49_dict_to_v50_dict(cls, states_dict):
+                answer_group['outcome']['dest_if_really_stuck'] = None
+
+            if state_dict['interaction']['default_outcome'] is not None:
+                state_dict['interaction']['default_outcome']['dest_if_really_stuck'] = None # pylint: disable=line-too-long
```
```suggestion
                state_dict[
                    'interaction']['default_outcome']['dest_if_really_stuck'] = None
```

Manan-Rathi replied:
Fixed.

iamprayush commented on core/domain/exp_domain.py:
```diff
@@ -2191,6 +2191,30 @@ def _convert_states_v49_dict_to_v50_dict(cls, states_dict):
+    def _convert_states_v50_dict_to_v51_dict(cls, states_dict):
+        """Converts from version 50 to 51. Version 51 adds
+        a new dest_if_really_stuck field to the Outcome class to redirect
+        the learners to a state for strengthening concepts.
```
```suggestion
        the learners to a state for strengthening concepts when they get really stuck.
```

Ditto here and elsewhere.

Manan-Rathi replied:
Fixed.

iamprayush commented on core/domain/exp_domain.py:
```diff
@@ -2434,6 +2458,28 @@ def _convert_v54_dict_to_v55_dict(cls, exploration_dict):
+
+        Args:
+            exploration_dict: dict. The dict representation of an exploration
+                with schema version v54.
```
*55

Manan-Rathi replied:
Fixed.

iamprayush commented on core/domain/exp_domain.py:
```diff
@@ -2434,6 +2458,28 @@ def _convert_v54_dict_to_v55_dict(cls, exploration_dict):
+
+        Returns:
+            dict. The dict representation of the Exploration domain object,
+            following schema version v55.
```
56

Manan-Rathi replied:
Fixed.

iamprayush commented on core/domain/question_domain.py:
```diff
@@ -1236,6 +1236,30 @@ def _convert_state_v49_dict_to_v50_dict(cls, question_state_dict):
 
+    @classmethod
+    def _convert_state_v50_dict_to_v51_dict(cls, question_state_dict):
+        """Converts from version 49 to 50. Version 50 adds
```
*50 to 51

Manan-Rathi replied:
Fixed.

iamprayush commented on core/domain/question_domain.py:
```diff
@@ -1236,6 +1236,30 @@ def _convert_state_v49_dict_to_v50_dict(cls, question_state_dict):
+            answer_group['outcome']['dest_if_really_stuck'] = None
+
+        if question_state_dict['interaction']['default_outcome'] is not None:
+            question_state_dict['interaction']['default_outcome']['dest_if_really_stuck'] = None # pylint: disable=line-too-long
```
```suggestion
            question_state_dict[
                'interaction']['default_outcome']['dest_if_really_stuck'] = None
```

Manan-Rathi replied:
Fixed.

iamprayush commented on index.yaml:
```diff
@@ -7,8 +7,9 @@ indexes:
 # manually, move them above the marker line.  The index.yaml file is
-# automatically uploaded to the admin console when you next deploy
-# your application using appcfg.py.
+# automatically uploaded to Cloud Datastore when you next deploy
```
Are changes in this file intended?

Manan-Rathi replied:
Didn't make these changes actually, might have been due to a pull.

iamprayush replied:
These are still there. Please sync with develop if needed, these shouldn't be here.

Manan-Rathi replied:
Removed these changes.

iamprayush commented on scripts/backend_tests_incomplete_coverage.txt:
```diff
@@ -26,6 +26,7 @@ core.storage.base_model.gae_models_test
 core.schema_utils_test
 core.domain.state_domain_test
 core.domain.learner_progress_services_test
+core.domain.question_fetchers_test
```
Why this change?

Manan-Rathi replied:
Since this file has 100% coverage, removed it now.

iamprayush reviewed (changes requested):
Took a pass, PTAL @Manan-Rathi, thanks!

Manan-Rathi committed:
Req changes

Manan-Rathi committed:
Fix merge conflicts

Manan-Rathi committed:
Minor edits

Manan-Rathi committed:
fixe

iamprayush commented:
Backend tests seem to be failing, PTAL @Manan-Rathi, thanks!

Manan-Rathi committed:
nit

Manan-Rathi committed:
Merge branch 'develop' of https://github.com/oppia/oppia into gsoc-sp1.1

Manan-Rathi committed:
nit

Manan-Rathi committed:
fix

Manan-Rathi commented:
All tests passing now, PTAL @iamprayush. Thanks!

iamprayush commented on core/domain/exp_domain.py:
```diff
@@ -2447,6 +2472,28 @@ def _convert_v54_dict_to_v55_dict(cls, exploration_dict):
+    def _convert_v55_dict_to_v56_dict(cls, exploration_dict):
+        """Converts a v55 exploration dict into a v56 exploration dict.
+        Version 56 adds a new dest_if_really_stuck field to the Outcome class
+        to redirect the learners to a state for strengthening concepts.
```
```suggestion
        to redirect the learners to a state for strengthening concepts when they get really stuck.
```
Please fix this here and elsewhere.

Manan-Rathi replied:
Done. Fixed everywhere.

iamprayush reviewed (approved):
Just a couple of nits, else LGTM!

seanlip reviewed (changes requested):
LGTM for codeowner files, but I don't think you've added the necessary validation in validate() for the new field?

Manan-Rathi committed:
Minor requested nits

Manan-Rathi commented:
> LGTM for codeowner files, but I don't think you've added the necessary validation in validate() for the new field?

Added the necessary validation in validate() of Outcome class. PTAL!

seanlip commented on core/domain/state_domain.py:
```diff
@@ -1595,6 +1604,12 @@ def validate(self):
         """
         self.feedback.validate()
 
+        if self.dest_if_really_stuck is not None:
```
Hm, don't you need more validation than this (and corresponding backend tests for the various cases)? E.g.:

- The dest_if_really_stuck should actually be a valid state name.
- If the answer group is marked as "correct" then there shouldn't be a dest_if_really_stuck
- The dest_if_really_stuck shouldn't match the name of the state it is currently being used for
- When checking that every state is reachable from the start, you'd need to consider the dest_if_really_stuck links as well (this is in addition to what I imagine are existing checks to ensure that, with only the regular dest links, every state can reach the end).

I might have missed some, so please think through the different cases and ensure there are tests for these. Better to make this robust now so that you don't get people creating invalid explorations which we need to clean up later.

/cc @iamprayush

Manan-Rathi replied:
Right, all the arguments make sense. Actually:

- A regex check would ensure that the state name entered is a valid one. Also, since the creator chooses the dest_if_really_stuck state from a dropdown or creates a new one, only valid states will be assigned to this field.
- In M1.2, a check is added that the dest_if_really_stuck UI become visible only if the answer group is not marked as correct.
- The same state name wouldn't appear in the dropdown. So, the case of a self-loop would be checked in the frontend itself.
- Actually, the check for this can only be implemented in M1.3 when new types of links would be created to connect the dest_if_really_stuck state.

Basically, these conditions will be applied in the frontend itself.
PS: I'll push the backend tests for the validation soon.

seanlip replied:
Note that all the validation checks have to be complete in the backend though. The backend should be robust to any data coming through the handlers, because it's technically possible for data to come to the backend handler from somewhere other than the frontend (e.g. a rogue POST request).

Manan-Rathi replied:
Added the validation checks to ensure that dest_if_really_stuck (in both outcome and default_outcome):
1. is a string if not null
2. should not be present when labelled_as_correct is true.
3. is a valid state
Also, added tests for the same. The fourth point is only possible in M1.3. PTAL @seanlip

Manan-Rathi committed:
Added validation tests.

Manan-Rathi committed:
fix

Manan-Rathi committed:
nit

seanlip reviewed (changes requested):
Thanks @Manan-Rathi, the validation looks good.

I have one question. In questions, we need to validate that the dest_if_really_stuck field is always None. Could you please do this as well?

And again -- are there other checks that should be added here?

Manan-Rathi commented:
> Thanks @Manan-Rathi, the validation looks good.
> 
> I have one question. In questions, we need to validate that the dest_if_really_stuck field is always None. Could you please do this as well?
> 
> And again -- are there other checks that should be added here?

Just so I get this right, do you mean that we should add a check to ensure that dest_if_really_stuck is initially null always?

seanlip commented:
No. I mean that you've put the validation checks in the exp_domain file since the checks in state_domain would apply to both Exploration and Question objects. This is fine, but then you'd need to also apply the relevant validation for Question objects (which also use States internally), which is that the dest_if_really_stuck field should be None.

Manan-Rathi committed:
Fixed merge conflicts

Manan-Rathi committed:
test fixes

Manan-Rathi committed:
Added validation for Question

Manan-Rathi committed:
lint fix

Manan-Rathi commented:
> No. I mean that you've put the validation checks in the exp_domain file since the checks in state_domain would apply to both Exploration and Question objects. This is fine, but then you'd need to also apply the relevant validation for Question objects (which also use States internally), which is that the dest_if_really_stuck field should be None.

Added the same. PTAL!

vojtechjelinek commented on core/domain/exp_domain.py:
```diff
@@ -1043,6 +1052,15 @@ def validate(self, strict=False):
 
+                default_outcome = state.interaction.default_outcome
+                if default_outcome.dest_if_really_stuck is not None:
+                    if not isinstance(
+                        default_outcome.dest_if_really_stuck, str):
```
```suggestion
                    if not isinstance(
                        default_outcome.dest_if_really_stuck, str
                    ):
```

Manan-Rathi replied:
Done.

vojtechjelinek commented on core/domain/exp_domain.py:
```diff
@@ -1138,6 +1156,20 @@ def validate(self, strict=False):
                         % default_outcome.dest)
 
+                # Check default if-stuck destinations.
+                if (default_outcome.dest_if_really_stuck is not None and
+                        default_outcome.dest_if_really_stuck not
+                            in all_state_names):
```
```suggestion
                if (
                    default_outcome.dest_if_really_stuck is not None and
                    default_outcome.dest_if_really_stuck not in all_state_names
                ):
```

vojtechjelinek replied:
ditto below

Manan-Rathi replied:
Done. PTAL!

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

Manan-Rathi committed:
changes

vojtechjelinek commented on core/domain/exp_domain.py:
```diff
@@ -1153,10 +1189,30 @@ def validate(self, strict=False):
+                    group.outcome.dest_if_really_stuck is not None and
+                    group.outcome.dest_if_really_stuck == state_name
+                ):
+                    raise utils.ValidationError(
+                            'The destination for a stuck learner cannot be the '
+                            'same state.')
```
```suggestion
                    raise utils.ValidationError(
                        'The destination for a stuck learner cannot be the '
                        'same state.')
```

Manan-Rathi replied:
Done!

vojtechjelinek reviewed (approved):
Thanks! LGTM! Just one small nit.

seanlip commented:
Hi @Manan-Rathi, the questions check looks fine. But it looks like you keep missing my other question.

> Are there other checks that should be added here?

Manan-Rathi commented:
> Hi @Manan-Rathi, the questions check looks fine. But it looks like you keep missing my other question.
> 
> > Are there other checks that should be added here?

Ya, I think we have covered all required validation checks for this field.

seanlip reviewed (approved):
OK, please make sure to add any additional checks that are needed in the future, to avoid data becoming corrupted in the backend. (This also applies to the check you postponed to a future PR.)

Otherwise, LGTM for codeowners!

Manan-Rathi commented:
> OK, please make sure to add any additional checks that are needed in the future, to avoid data becoming corrupted in the backend. (This also applies to the check you postponed to a future PR.)
> 
> Otherwise, LGTM for codeowners!

Ya, will ensure that.

krishnarao22 commented on core/tests/protractor_desktop/explorationHistoryTab.js:
```diff
@@ -138,94 +138,98 @@ describe('Exploration history', function() {
       },
       12: {
-        text: '    feedback:',
+        text: '    dest_if_really_stuck: null',
```
Why are we adding `dest_if_really_stuck` in field number 12 specifically? Why can we not just add it at the end?

Manan-Rathi replied:
Just to preserve the general order we've maintained everywhere. For eg, line 477 in the same file.

krishnarao22 reviewed (commented):
Just one question @Manan-Rathi

gp201 reviewed (approved):
LGTM

Manan-Rathi committed:
Fix merge conflicts

Manan-Rathi committed:
Added missing field

Manan-Rathi committed:
fixes

Manan-Rathi committed:
nits

Manan-Rathi committed:
fixes

Manan-Rathi committed:
nit

kevintab95 reviewed (approved):
Thanks @Manan-Rathi, LGTM for codeowner files. Sorry for the delay!

Manan-Rathi committed:
Fix Merge conflicts

Manan-Rathi commented:
Answered your comment, PTAL @krishnarao22! Also, the CI checks fail due to flakes from a recent pull, will pull again once it's reverted.

krishnarao22 reviewed (approved):
LGTM for 1 code owner file, thanks @Manan-Rathi!

Manan-Rathi committed:
Merge branch 'develop' of https://github.com/oppia/oppia into gsoc-sp1.1

Manan-Rathi commented:
There seems to be some frontend coverage issue in conversation.skin.component.ts. I've made no changes to the learner side but karma CI checks still fail mentioning incomplete coverage in the file. Here is a link to the [PR Run](https://github.com/oppia/oppia/runs/7355878581?check_suite_focus=true). Here is a link to another PR's [run](https://github.com/oppia/oppia/runs/7353434793?check_suite_focus=true) that faces the same issue. I'm not really sure what's wrong. I've added it to the NOT_FULLY_COVERED_FILES list. Thanks!


However, when I add the file, it says that the file's fully covered. And so the tests fail.

Removed it, and pushed again. Tests pass now. Was a flake probably!

Manan-Rathi committed:
Added conversation-skin.component.ts to NOT_FULLY_COVERED

Manan-Rathi committed:
Minor nit

Manan-Rathi committed:
Removed the file
